lib/Feature/FeatureModelParser.cpp

```diff
@@ -233,4 +234,318 @@ std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> FeatureModelXmlParser::parseDoc() {
+      // The first character has to be a colon followed by the type of
+      // the feature (m for mandatory, o for optional, a for alternative)
+      std::string::size_type Pos = CurrentIndentationLevel * Indentation.length() + 2;
+      std::optional<std::tuple<int, int>> Cardinalities;
+
+      switch (To.at(Pos - 1)) {
+      case 'o':
+        // Code for optional
+        Opt = true;
+        break;
+      case 'g':
+        // Code for an or group with different cardinalities
+        Opt = false;
+        // Extract the cardinality
+        Cardinalities = extractCardinality(To);
+        if (!Cardinalities.has_value()) {
+          return false;
+        }
+        break;
+      case ' ':
+        // Code for alternative child
+        Pos--;
+        break;
+      }
+      // Extract the name
+      Name = ToStringRef.substr(Pos + 1, ToStringRef.find(' ', Pos + 1) - Pos - 1).str();
+
+      // Remove the cardinality
+      if (Name.find_first_of('[') != std::string::npos) {
+        Name = ToStringRef.substr(Pos + 1, ToStringRef.find('[', Pos + 1) - Pos - 1).str();
+
+        if (Name.empty()) {
+          // In this case, the name could also be after the cardinality.
+          // According to the examples provided by S.P.L.O.T., this is a valid format.
+          auto Remainder =
+              ToStringRef.substr(Pos + 1, ToStringRef.size() - Pos - 1);
+          auto Tokens = Remainder.split(' ');
+          if (!Tokens.second.empty()) {
+            Name = Tokens.second.str();
+          }
+        }
+      }
+
+      // Note that we ignore the ID and use the name of the feature
+      // as unique identifier.
+      if (Name.find_first_of('(') != std::string::npos) {
+        Name = Name.substr(0, Name.find('(', 0));
+      }
+
+      // If there is no name, provide an artificial one
+      if (Name.empty()) {
+        OrGroupCounter++;
+        Name = "group_" + std::to_string(OrGroupCounter);
+      }
+
+      // Create the feature
+      FMB.makeFeature<BinaryFeature>(Name, Opt);
+      IndentationToParentMapping[CurrentIndentationLevel] = Name;
+
+      // Add parent from the upper indentation level if there is one
+      if (LastIndentationLevel != -1 && CurrentIndentationLevel == RootIndentation) {
+        std::cerr << "Only one feature can be root and have the same indentation as root." << std::endl;
+        return false;
+      }
+
+      if (LastIndentationLevel != -1) {
+        auto Parent = IndentationToParentMapping.find(CurrentIndentationLevel - 1);
+        assert(Parent != IndentationToParentMapping.end());
+        FMB.addParent(Name, Parent->second);
+      }
+
+      // Add the or group to the feature model if it is completely parsed
+      auto OrGroup = OrGroupMapping.find(CurrentIndentationLevel);
+      if (OrGroup != OrGroupMapping.end()) {
+        FMB.emplaceRelationship(std::get<1>(OrGroup->second), std::get<2>(OrGroup->second), std::get<0>(OrGroup->second));
```

Comment:
formatting, this line is obviously to to long. If you need help setting up clang-format just ping me